oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/OptimizerController.kt

```diff
@@ -0,0 +1,230 @@
+        val name: String,
+        val type: Type,
+        val value : Double? = null,
+        val lb: Double? = null,
+        val ub: Double? = null
+    )
+
+    fun buildTree(config: OptimizerEndpoint.Simulation): TreeItem<Parameter> {
+        val root = TreeItem<Parameter>(Parameter("root", Type.Root))
+        val root1 = TreeItem<Parameter>(Parameter("Inputs", Type.Root))
+        val root2 = TreeItem<Parameter>(Parameter("Outputs", Type.Root))
+
+        val inputs: List<TreeItem<Parameter>> = config.inputs.map { it ->
+            TreeItem(Parameter(it.name, Type.Input, it.currentValue, it.lowerBound, it.upperBound))
+        }
+        val outputs: List<TreeItem<Parameter>> = config.outputs.map { it ->
+            TreeItem(Parameter(it.name, Type.Input))
+        }
+
+        root1.isExpanded = true
+        root2.isExpanded = true
+
+        root1.children.addAll(inputs)
+        root2.children.addAll(outputs)
+
+        root.children.addAll(root1, root2)
+        return root
+    }
+
+    @FXML fun initialize() {
+        senderColumn.setCellValueFactory { dataFeatures -> SimpleStringProperty(dataFeatures.value.sender) }
+        timeColumn.setCellValueFactory { dataFeatures -> SimpleStringProperty(dataFeatures.value.receiveTime.toString()) }
+        messageColumn.setCellValueFactory { dataFeatures -> SimpleStringProperty(dataFeatures.value.message) }
+
+        senderColumn1.setCellValueFactory { dataFeatures -> SimpleStringProperty(dataFeatures.value.name) }
+        typeColumn.setCellValueFactory { dataFeatures -> SimpleStringProperty(dataFeatures.value.resultType) }
+        inputColumn.setCellValueFactory { dataFeatures -> SimpleStringProperty(dataFeatures.value.inputs) }
+        outputColumn.setCellValueFactory { dataFeatures -> SimpleStringProperty(dataFeatures.value.outputs) }
+
+        nameColumn.cellValueFactory = TreeItemPropertyValueFactory<Parameter, String>("name")
+        valueColumn.cellValueFactory = TreeItemPropertyValueFactory<Parameter, String>("value")
+        lbColumn.cellValueFactory = TreeItemPropertyValueFactory<Parameter, String>("lb")
+        upColumn.cellValueFactory = TreeItemPropertyValueFactory<Parameter, String>("ub")
+
+        nodesList.selectionModel.selectedItemProperty().addListener { src, oldV, newV -> showNode(newV) }
+
+        timeOutTextField.setOnKeyPressed{ event ->
+            if(event.code == KeyCode.ESCAPE){
+                //discard
+                timeOutTextField.text = endpoint.simulationsByName.getValue(nodesList.selectedItem!!).timeOut!!.toMillis().toString()
+                view.requestFocus()
+            }
+            else if (event.code == KeyCode.ENTER) {
+                //commit
+                try{
+                    endpoint.setDuration(nodesList.selectedItem, Duration.ofMillis(timeOutTextField.text.toLong()))
+                }
+                catch (e : NumberFormatException){
+                    timeOutTextField.text = ""
+                }
```

Comment:
Oook, theres a lot we can do with this:

1. since you have control of `endpoint: OptimizerEndpoint`, you could make `duration` a simple property
2. I would strongly encourage you to use `toLongOrNull()`, which is new in kotlin for strings, and allows you to avoid exception flow

then we get something like

```kt
val durationMillis = timeoutTextField.text.toLongOrNull()
if(durationMillis == null){
  timeOutTextField.text = "";
}
else {
  endpoint.duration = Duration.ofMillis(timeoutMillis)  
}
```

Reply:
changed